src/gravel/controllers/kv.py

```diff
@@ -12,125 +12,395 @@
 # GNU General Public License for more details.
 
 import asyncio
+
+# The other obvious choice here would be leveldb, given that's
+# already used by ceph, but dbm is already approximately
+# part of python core, so once less dependency.
+import dbm
+import sys
+import threading
 from logging import Logger
+from pathlib import Path
 from typing import Callable, List, Optional
 
-import aetcd3
-import aetcd3.events
-import aetcd3.locks
-import grpclib.exceptions
 from fastapi.logger import logger as fastapi_logger
 
 logger: Logger = fastapi_logger
 
+# Liberated from gravel/controllers/orch/ceph.py
+# TODO: make this common / combine somehow?
+try:
+    import rados
+except ModuleNotFoundError:
+    pass
 
-class Lock:
-    _lock: aetcd3.locks.Lock
-    _is_acquired: bool
-
-    def __init__(self, lock: aetcd3.locks.Lock):
-        self._lock = lock
-        self._is_acquired = False
 
-    async def acquire(self) -> None:
-        await self._lock.acquire()
-        self._is_acquired = True
-
-    async def release(self) -> None:
-        if not self._is_acquired:
-            return
-        await self._lock.release()
-        self._is_acquired = False
+# Liberated from gravel/controllers/orch/ceph.py
+# TODO: make this common / combine somehow?
+class MissingSystemDependency(Exception):
+    pass
 
 
 class KV:
 
-    _client: Optional[aetcd3.Etcd3Client]
-    _is_open: bool
-    _is_closing: bool
+    # TODO: properly declare types of class variables (_db, _cluster, etc.)
 
     def __init__(self):
-        self._client = None
-        self._is_open = False
-        self._is_closing = False
+        if "rados" not in sys.modules:
+            raise MissingSystemDependency("python3-rados module not found")
+        # TODO: this should be created somewhere else (src/gravel/controllers/config.py?)
+        var_lib_aquarium: Path = Path("/var/lib/aquarium")
+        if not var_lib_aquarium.exists():
+            var_lib_aquarium.mkdir(0o700)
+        # this will fail with "_gdbm.error: [Errno 11] Resource temporarily unavailable: '/var/lib/aquarium/kvstore'
+        # if someone else has it open ... need KV to be a singleton?
+        # Or, do we only open it when reading/writing in individual methods?
```

Comment:
Practically, we probably gain nothing from not keeping it open. I did spend a while experimenting with a couple different ways of opening/closing in individual R/W methods, but all that did was make the implementation irritatingly complex. The only reason this came up at all was when I was doing some manual testing while aquarium was running, I wanted to see what was in the cache, and discovered I couldn't use `gdbmtool`, or python interactively, to poke around because the file was locked.

Reply:
I've updated the comment appropriately